Logstash on Docker

krallin opened the pull request:
Still a work in progress (namely, tests are missing), but there are a few things I'd like to start getting thoughts on:
- How do we want users to install Logstash plugins? Currently, there's a file they must edit and commit. I think that's a reasonable approach, as it lets users take advantage of Docker caching. The alternative would be to provide plugins to install through an environment variable, but that would slow down boot times.
- Do we want to have a set of default plugins? Currently, there are none. I'm not familiar enough with the Logstash ecosystem to be know what plugins should be included off the top of my head, but I think it'd be worth researching this a little bit.

Cheers,

cc @fancyremarker @aaw

krallin committed:
Logstash on Docker

krallin committed:
Add Logstash tests

krallin commented:
Added tests. 

@aaw , @fancyremarker , this should be ready for review now.

Note: I reused the same names we use in Gentlemanjerry for environment variables (`LOGSTASH_FILTERS` and `LOGSTASH_OUTPUT_CONFIG`); do we want to consider using slightly different and more consistent names? (e.g. (`LOGSTASH_FILTER_CONFIG`, `LOGSTASH_OUTPUT_CONFIG`)

Cheers,

krallin committed:
Give Logstash more time to come online in tests

krallin commented:
@fancyremarker ,

Looks like tests can take a while to run (at least on a low-powered staging environment), so for deployment purposes, I'd need to increase (substantially) the timeouts in tests (e.g. to several minutes), which in turn makes them essentially useless in development.

That got me thinking, though. Do we want all customers that deploy this image to have to run the tests incorporated in the Docker build to begin with? If not, perhaps we can find a way to only run in dev / test.

Cheers,

krallin committed:
Warn against using stdout output, remove default

krallin committed:
Don't wrap in useless bash

fancyremarker commented:
@krallin: I'm fine with not running tests in the default `docker build`, if that's more practical. I can think of two ways to run tests outside of the default `docker build`:
- Include 2 Dockerfiles, one for deployment and one for testing, and only run the Bats test suite in the test Dockerfile.
- Move the tests out of a Bats suite inside the image, and into a suite that runs outside the image, executing tests against containers launched from the image. (This would involve more rewriting.)

aaw commented:
Looks great! 

A default set of plugins sounds like a good idea and I'd vote for including more rather than less in this image just to make setup a little easier (e.g., just add all output plugins and maybe our syslog plugin). I don't feel strongly about including a ton of output plugins, though, so I'm totally fine if you want to go with allowing them to edit an initially empty plugins file instead. Configuring them through environment variables is probably not a good idea, since there's a chance an app with a lot of plugins would time out our 5 minute health check.

Renaming the environment variables to `LOGSTASH_FILTER_CONFIG` and `LOGSTASH_OUTPUT_CONFIG` is totally fine.

krallin committed:
Rename LOGSTASH_FILTERS to LOGSTASH_FILTER_CONFIG

krallin committed:
Edit Gemfile directly, include syslog gem

Simpler than having a separate install-plugins mechanism, and makes it
much easier for us to install our customized syslog gem.

krallin commented:
All right:
- Renamed `LOGSTASH_FILTERS` to `LOGSTASH_FILTER_CONFIG` for consistency.
- Included our syslog gem. Figured I'd just let the users edit the `Gemfile` directly, though this means we might have to be a bit more careful when updating to a newer Logstash (e.g. if some of the hardcoded dependencies have changed).

Cheers,

krallin committed:
Only run tests as part of build process

This speeds up deployment on Aptible (where users aren't expected to
need to run tests prior to deploying), all the while ensuring tests do
run when executing `make build`.

krallin commented:
Updated to only run tests on `make build` with a bit of `sed`.

krallin commented:
This should be ready to go now!

fancyremarker commented:
Looks great! Merging. I added a [Quay repo](https://quay.io/repository/aptible/logstash) since it's referenced in the README, but I haven't pushed this built image there yet. (We may not actually need the Quay repo at all.)
